File: lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/stack/topbar/TopBarController.java
package com.reactnativenavigation.viewcontrollers.stack.topbar;

import android.animation.Animator;
import android.content.Context;
import android.graphics.drawable.Drawable;
import android.view.MenuItem;
import android.view.View;

import com.reactnativenavigation.options.AnimationOptions;
import com.reactnativenavigation.options.Options;
import com.reactnativenavigation.options.animations.ViewAnimationOptions;
import com.reactnativenavigation.viewcontrollers.stack.topbar.button.ButtonController;
import com.reactnativenavigation.viewcontrollers.stack.topbar.title.TitleBarReactViewController;
import com.reactnativenavigation.views.stack.StackLayout;
import com.reactnativenavigation.views.stack.topbar.TopBar;
import com.reactnativenavigation.views.stack.topbar.titlebar.TitleBar;

import java.util.List;

import androidx.annotation.VisibleForTesting;
import androidx.viewpager.widget.ViewPager;

import static com.reactnativenavigation.utils.CollectionUtils.*;
import static com.reactnativenavigation.utils.ObjectUtils.perform;
import static com.reactnativenavigation.utils.ViewUtils.isVisible;


public class TopBarController {
    private TopBar topBar;
    private TitleBar titleBar;
    private TopBarAnimator animator;

    public TopBarAnimator getAnimator() {
        return animator;
    }

    public MenuItem getRightButton(int index) {
        return titleBar.getRightButton(index);
    }

    public TopBar getView() {
        return topBar;
    }

    public int getHeight() {
        return perform(topBar, 0, View::getHeight);
    }

    public int getRightButtonsCount() {
        return topBar.getRightButtonsCount();
    }

    public Drawable getLeftButton() {
        return titleBar.getNavigationIcon();
    }

    @VisibleForTesting
    public void setAnimator(TopBarAnimator animator) {
        this.animator = animator;
    }

    public TopBarController() {
        animator = new TopBarAnimator();
    }

    public TopBar createView(Context context, StackLayout parent) {
        if (topBar == null) {
            topBar = createTopBar(context, parent);
            titleBar = topBar.getTitleBar();
            animator.bindView(topBar, parent);
        }
        return topBar;
    }

    protected TopBar createTopBar(Context context, StackLayout stackLayout) {
        return new TopBar(context);
    }

    public void initTopTabs(ViewPager viewPager) {
        topBar.initTopTabs(viewPager);
    }

    public void clearTopTabs() {
        topBar.clearTopTabs();
    }

    public Animator getPushAnimation(Options appearingOptions, ViewAnimationOptions topBar) {
        return animator.getPushAnimation(appearingOptions, topBar);
    }

    public Animator getPopAnimation(Options appearingOptions, ViewAnimationOptions topBar) {
        return animator.getPopAnimation(appearingOptions, topBar);
    }

    public void show() {
        if (isVisible(topBar) || animator.isAnimatingShow()) return;
        topBar.setVisibility(View.VISIBLE);
    }

    public void showAnimate(AnimationOptions options, int translationDy) {
        if (isVisible(topBar) || animator.isAnimatingShow()) return;
        animator.show(options, translationDy);
    }

    public void hide() {
        if (!animator.isAnimatingHide()) topBar.setVisibility(View.GONE);
    }

    public void hideAnimate(AnimationOptions options, float translationStart, float translationEnd) {
        hideAnimate(options, translationStart, translationEnd, () -> {});
    }

    private void hideAnimate(AnimationOptions options, float translationStart, float translationEnd, Runnable onAnimationEnd) {
        if (!isVisible(topBar) || animator.isAnimatingHide()) return;
        animator.hide(options, translationStart, translationEnd, onAnimationEnd);
    }

    public void setTitleComponent(TitleBarReactViewController component) {
        topBar.setTitleComponent(component.getView());
    }

    public void applyRightButtons(List<ButtonController> toAdd) {
        topBar.clearRightButtons();
        forEachIndexed(toAdd, (b, i) -> b.addToMenu(titleBar, (toAdd.size() - i) * 10));
    }

    public void mergeRightButtons(List<ButtonController> toAdd, List<ButtonController> toRemove) {
        forEach(toRemove, btn -> topBar.removeRightButton(btn));
        forEachIndexed(toAdd, (b, i) -> b.addToMenu(titleBar, (toAdd.size() - i) * 10));
    }

    public void setLeftButtons(List<ButtonController> leftButtons) {
        titleBar.setLeftButtons(leftButtons);
    }
}

File: lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/stack/topbar/TopBarController.kt
package com.reactnativenavigation.viewcontrollers.stack.topbar

import android.animation.Animator
import android.content.Context
import android.graphics.drawable.Drawable
import android.view.MenuItem
import android.view.View
import androidx.annotation.VisibleForTesting
import androidx.viewpager.widget.ViewPager
import com.reactnativenavigation.options.AnimationOptions
import com.reactnativenavigation.options.Options
import com.reactnativenavigation.utils.CollectionUtils
import com.reactnativenavigation.utils.ViewUtils
import com.reactnativenavigation.viewcontrollers.stack.topbar.button.ButtonController
import com.reactnativenavigation.viewcontrollers.stack.topbar.title.TitleBarReactViewController
import com.reactnativenavigation.views.stack.StackLayout
import com.reactnativenavigation.views.stack.topbar.TopBar
import com.reactnativenavigation.views.stack.topbar.titlebar.TitleBar

open class TopBarController {
    lateinit var view: TopBar
    private lateinit var titleBar: TitleBar
    @VisibleForTesting var animator: TopBarAnimator = TopBarAnimator()

    val height: Int
        get() = view.height
    val rightButtonsCount: Int
        get() = view.rightButtonsCount
    val leftButton: Drawable?
        get() = titleBar.navigationIcon

    fun getRightButton(index: Int): MenuItem = titleBar.getRightButton(index)

    fun createView(context: Context, parent: StackLayout): TopBar {
        if (!::view.isInitialized) {
            view = createTopBar(context, parent)
            titleBar = view.titleBar
            animator.bindView(view)
        }
        return view
    }

    protected open fun createTopBar(context: Context, stackLayout: StackLayout): TopBar {
        return TopBar(context)
    }

    fun initTopTabs(viewPager: ViewPager?) = view.initTopTabs(viewPager)

    fun clearTopTabs() = view.clearTopTabs()

    fun getPushAnimation(
            appearingOptions: Options,
            additionalDy: Float = 0f
    ): Animator {
        return animator.getPushAnimation(
                appearingOptions.animations.push.topBar,
                appearingOptions.topBar.visible,
                additionalDy
        )
    }

    fun getPopAnimation(appearingOptions: Options, disappearingOptions: Options): Animator {
        return animator.getPopAnimation(disappearingOptions.animations.pop.topBar, appearingOptions.topBar.visible)
    }

    fun show() {
        if (ViewUtils.isVisible(view) || animator.isAnimatingShow()) return
        view.visibility = View.VISIBLE
    }

    fun showAnimate(options: AnimationOptions, additionalDy: Float) {
        if (ViewUtils.isVisible(view) || animator.isAnimatingShow()) return
        animator.show(options, additionalDy)
    }

    fun hide() {
        if (!animator.isAnimatingHide()) view.visibility = View.GONE
    }

    fun hideAnimate(options: AnimationOptions, additionalDy: Float) {
        if (!ViewUtils.isVisible(view) || animator.isAnimatingHide()) return
        animator.hide(options, additionalDy)
    }

    fun setTitleComponent(component: TitleBarReactViewController) {
        view.setTitleComponent(component.view)
    }

    fun applyRightButtons(toAdd: List<ButtonController>) {
        view.clearRightButtons()
        CollectionUtils.forEachIndexed(toAdd) { b: ButtonController, i: Int -> b.addToMenu(titleBar, (toAdd.size - i) * 10) }
    }

    fun mergeRightButtons(toAdd: List<ButtonController>, toRemove: List<ButtonController>?) {
        CollectionUtils.forEach(toRemove) { btn: ButtonController? -> view.removeRightButton(btn) }
        CollectionUtils.forEachIndexed(toAdd) { b: ButtonController, i: Int -> b.addToMenu(titleBar, (toAdd.size - i) * 10) }
    }

    fun setLeftButtons(leftButtons: List<ButtonController?>?) {
        titleBar.setLeftButtons(leftButtons)
    }

}